### PR TITLE
Bug years as dates

### PR DESCRIPTION
**Pull request by jsandoe**
Fixes the bugs that Jacob reported involving the interpretation of dates in the plugin. When the time attribute contained years, the plugin interpreted them as numbers, even when the attribute type is "date/time". This meant that the tracking global on the CODAP side would be confused.

There was code to check the attribute type, but it was poorly implemented (by me) and broken.

Other changes in this PR:
* Add support for "prettier" code formatting
* Add the beginnings of a README.md
* Fix bugs to better to respond to events on the CODAP side. Sometimes the plugin would not do the right thing when a new dataset was added in CODAP or cases were added to an existing dataset.
* Temporarily revert the webpack build of the Info plugin pending a better understanding of webpack configuration. Somehow the Info Plugin's javascript crept into the Sonify plugin.
* Add a source map to the build.

### CODE REVIEW

File: Sonify/README.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+- Open CODAP in one tab of a browser.
+- Open the Sonify plugin in another tab.
+- Drag the URL for the latter tab into the tab of the former.
+  Wait for the CODAP tab to become active, then drop the URL in the CODAP
+  workspace.
+
+Alternatively, if CODAP's URL is http://localhost:8080/CODAP/ and the plugin's
+URL is http://localhost:8080/Sonify/, then, invoke them together this way:
+
+    `http://localhost:8080/CODAP/?di=http://localhost:8080/Sonify/`
```
Comment (ttsuchiya) [inline Sonify/README.md:34]: Thank you for the README! 

Also, for testing the plugin with a release version of CODAP on `https`, I did the following:
- Install [mkcert](https://github.com/FiloSottile/mkcert). On Mac: `brew install mkcert`
- Create and install the CA in KeyChain with `mkcert -install`
- `cd` to this directory.
- `mkcert -key-file key.pem -cert-file cert.pem localhost`
- `npx http-server . -S` => Go to `https://localhost:8080/Sonify/`

File: Sonify/package.json [hunk 2]
```diff
@@ -11,14 +11,15 @@
     "build": "webpack --config webpack.config.js",
     "test": "echo \"No automated tests at this time. See test/test-cases.md.\" && exit 1"
   },
-  "author": "Takahiko Tsuchiya",
+  "author": "Concord Sonification Team",
```
Comment (ttsuchiya) [inline Sonify/package.json:14]: 👍

File: Sonify/src/CodapPluginHelper.js [hunk 3]
```diff
@@ -2,760 +2,904 @@
...
+      notice.resource.includes("dataContextChangeNotice")
+    ) {
+      // This is for checking repeated interleaving duplicates (e.g., deleteCollection & moveAttribute)
+      res = false;
+    } else if (this.prevNotice) {
+      res = JSON.stringify(notice) !== JSON.stringify(this.prevNotice);
+      this.prevDeleteNotice = null;
+    }
+
+    // Some nasty code for special-case nasty behaviors.
```
Comment (ttsuchiya) [inline Sonify/src/CodapPluginHelper.js:156]: 😓

**Review by ttsuchiya — APPROVED** [review]
It's a lot more readable after the code formatting! But could you point me to where the date-related bug fix was? I couldn't really find it in the code because the diff view is all red / green 😅

File: Sonify/README.md (hunk 1 above)
Comment (jsandoe) [reply Sonify/README.md:34]: I'll add that.

**Comment by jsandoe** [comment]
> It's a lot more readable after the code formatting! But could you point me to where the date-related bug fix was? I couldn't really find it in the code because the diff view is all red / green 😅

I think I should have made the code formatting a separate PR, before I made that change, it would have been obvious what each delta referred to.

The changes that actually fixed Jacob's bug were these:
* in `CodapHelper.js`, add a new property `this.contextStructures` which contains the raw dataContext metadata as returned by codap.
* in `CodapHelper.js`, modify `queryCollectionList` to fetch and populate the `contextStructures` object while extracting collection information.
* in `CodapHelper.js`, add the method `getAttributeDefsForContext(..)` to extract the full attribute definitions from `contextStructures`.
* in `app.js`, the body of `getAttributeType(..)` was rewritten to pull this information by calling `helper.getAttributeDefsForContext(..)`.
* in `app.js`, `setIfDateTimeAttribute(..)` was adjusted to make the correct call to `getAttributeType(..)`

Hope this helps.

Jonathan

**Comment by ttsuchiya** [comment]
Great, thanks Jonathan for the pointers and fixing these parts. These host / plugin contracts are quite intricate and fascinating to me!

**Pull request merged**